<a href="https://colab.research.google.com/github/alzimna/Portfolio/blob/main/Notebook%20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Overview**

Pada tahun 2020, dunia diguncang oleh suatu penyakit mematikan yang dinamakan Covid 19. Selain membuat imun penderitanya lemah, penyakit ini ternyata menyebabkan ekonomi dunia juga lesu. Hal ini disebabkan salah satunya karena para pelaku ekonomi berjaga-jaga dengan menyimpan uang cash sebanyak-banyaknya sehingga daya beli turun. Pada pasar saham, mayoritas investor akan menarik uangnya dari pasar saham dan membuat harga saham turun.

Akan dilihat apakah pandemi berpengaruh terhadap rata-rata EPS dan harga penutupan saham-saham pada indeks LQ45 sebelum pandemi dan setelah pandemi. Akan digunakan data EPS emiten-emiten pada indeks LQ45 pada tahun 2019 dan 2022, dan harga penutupan pada tanggal 30 Desember 2019 dan 30 Desember 2022.

# **Project Method**
Metode yang digunakan pada projek ini adalah materi Bab 5 pada buku "*Methods of Multivariate Analysis*", Oleh  ALVIN C. RENCHER yaitu menggunakan uji mean dari dua data. Naturally, jika dua data mempunyai mean yang berbeda maka dapat disimpulkan bahwa treatment yang diberikan pada masing-masing subjek mempunyai pengaruh yang signifikan. Pada projek ini digunakan bahasa pemrograman R.

Untuk kemudahan dimisalkan
\begin{align*}
    y_1, x_1 &= \text{EPS tahunan pada tahun 2022, dan 2019}\\
    y_2, x_2 &= \text{Harga pada akhir tahun 2022, dan 2019}
\end{align*}

Untuk membandingkan efek pandemi dan sebelum pandemi pada awalnya dicari $d_1$ dan $d_2$ yang berturut-turut merupakan selisih dari $y_1,x_1$ dan $y_2,x_2$.

# **Code Snippet**

In [1]:
install.packages('matlib')
library("matlib")
library(httr)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘rbibutils’, ‘cowplot’, ‘Deriv’, ‘microbenchmark’, ‘Rdpack’, ‘numDeriv’, ‘doBy’, ‘SparseM’, ‘MatrixModels’, ‘minqa’, ‘nloptr’, ‘reformulas’, ‘RcppEigen’, ‘carData’, ‘abind’, ‘Formula’, ‘pbkrtest’, ‘quantreg’, ‘lme4’, ‘rgl’, ‘car’




In [2]:
github_link <- "https://github.com/alzimna/Portfolio/raw/refs/heads/main/Test%20on%20Two%20Means/output.xlsx"
temp_file <- tempfile(fileext = ".xlsx")
req <- GET(github_link,
          # authenticate using GITHUB_PAT
           authenticate(Sys.getenv("GITHUB_PAT"), ""),
          # write result to disk
           write_disk(path = temp_file))
Y = readxl::read_excel(temp_file)[,2:5]
Y

EPS 2022 PRICE_2022 EPS 2019 PRICE_2019
1    38.83    496.0      59.58   1495.0000
2     0.08   3850.0       0.01   1555.0000
3   121.77   1400.0      36.06    790.0000
4    68.76   2650.0      26.79    880.0000
5   159.00   1985.0      77.47    840.0000
6     1.15   3720.0     -12.62   3100.0000
7   715.00   5700.0     536.00   6925.0000
8   330.00   8550.0     231.80   6685.0000
9   983.00   9225.0     585.00   7850.0000
10  338.00   4940.0     281.00   4400.0000
11  288.00   1350.0      17.77   2120.0000
12  441.26   4962.5     294.45   3837.5000
13  102.54   1290.0       7.43    330.0000
14    0.00    755.0       0.00   1503.6688
15  179.00   5650.0     222.00   6500.0000
16   89.39   1030.0     -26.88    557.5000
17    0.01    915.0       0.00    268.0000
18  105.00   2140.0      67.00   3150.0000
19    0.02   1620.0       0.00    264.0000
20  393.00  10000.0     432.00  11150.0000
21    0.02   7100.0       0.01   3640.0000
22  724.00   6725.0     559.00   7925.0000
23    0.09   2730.0       0.00   1195.0000
24    0.16   8725.0       0.05   7700.0000
25  529.44   9900.0     498.56  19025.0000
26    1.07  39025.0       0.12  11475.0000
27  122.00   1295.0     151.00   1535.0000
28   72.71   2090.0      53.48   1620.0000
29    0.00   4120.0       0.00   1070.0000
30    0.02   1015.0       0.00    732.2751
31    0.01   1760.0       0.00   2170.0000
32 1090.00   3690.0     371.00   2660.0000
33   13.39    206.0      14.58    282.0000
34   36.82    755.0      26.92    632.6702
35  397.00   6575.0     401.88  12000.0000
36  342.00   2530.0     543.60    724.0000
37   73.44   2300.0      39.26   1230.0000
38  140.00   1170.0     -82.00    825.0000
39  209.49   3750.0     251.13   3970.0000
40   69.00   1100.0      46.00    805.0000
41    0.00   2570.0       0.00   2593.7500
42 5680.00  26075.0    3030.00  21525.0000
43  141.00   4700.0     194.00   8400.0000

In [5]:
n = dim(Y)[1]
p = dim(Y)[2]/2
colnames(Y) = c("y1","y2","x1","x2")
Y$d1 = Y$y1-Y$x1
Y$d2 = Y$y2-Y$x2
Y = as.matrix(Y)
Y

y1      y2      x1      x2         d1      d2        
 [1,]   38.83   496.0   59.58  1495.0000  -20.75  -999.0000
 [2,]    0.08  3850.0    0.01  1555.0000    0.07  2295.0000
 [3,]  121.77  1400.0   36.06   790.0000   85.71   610.0000
 [4,]   68.76  2650.0   26.79   880.0000   41.97  1770.0000
 [5,]  159.00  1985.0   77.47   840.0000   81.53  1145.0000
 [6,]    1.15  3720.0  -12.62  3100.0000   13.77   620.0000
 [7,]  715.00  5700.0  536.00  6925.0000  179.00 -1225.0000
 [8,]  330.00  8550.0  231.80  6685.0000   98.20  1865.0000
 [9,]  983.00  9225.0  585.00  7850.0000  398.00  1375.0000
[10,]  338.00  4940.0  281.00  4400.0000   57.00   540.0000
[11,]  288.00  1350.0   17.77  2120.0000  270.23  -770.0000
[12,]  441.26  4962.5  294.45  3837.5000  146.81  1125.0000
[13,]  102.54  1290.0    7.43   330.0000   95.11   960.0000
[14,]    0.00   755.0    0.00  1503.6688    0.00  -748.6688
[15,]  179.00  5650.0  222.00  6500.0000  -43.00  -850.0000
[16,]   89.39  1030.0  -26.88   557.5000  116.27   472.5000
[17,]    0.01   915.0    0.00   268.0000    0.01   647.0000
[18,]  105.00  2140.0   67.00  3150.0000   38.00 -1010.0000
[19,]    0.02  1620.0    0.00   264.0000    0.02  1356.0000
[20,]  393.00 10000.0  432.00 11150.0000  -39.00 -1150.0000
[21,]    0.02  7100.0    0.01  3640.0000    0.01  3460.0000
[22,]  724.00  6725.0  559.00  7925.0000  165.00 -1200.0000
[23,]    0.09  2730.0    0.00  1195.0000    0.09  1535.0000
[24,]    0.16  8725.0    0.05  7700.0000    0.11  1025.0000
[25,]  529.44  9900.0  498.56 19025.0000   30.88 -9125.0000
[26,]    1.07 39025.0    0.12 11475.0000    0.95 27550.0000
[27,]  122.00  1295.0  151.00  1535.0000  -29.00  -240.0000
[28,]   72.71  2090.0   53.48  1620.0000   19.23   470.0000
[29,]    0.00  4120.0    0.00  1070.0000    0.00  3050.0000
[30,]    0.02  1015.0    0.00   732.2751    0.02   282.7249
[31,]    0.01  1760.0    0.00  2170.0000    0.01  -410.0000
[32,] 1090.00  3690.0  371.00  2660.0000  719.00  1030.0000
[33,]   13.39   206.0   14.58   282.0000   -1.19   -76.0000
[34,]   36.82   755.0   26.92   632.6702    9.90   122.3298
[35,]  397.00  6575.0  401.88 12000.0000   -4.88 -5425.0000
[36,]  342.00  2530.0  543.60   724.0000 -201.60  1806.0000
[37,]   73.44  2300.0   39.26  1230.0000   34.18  1070.0000
[38,]  140.00  1170.0  -82.00   825.0000  222.00   345.0000
[39,]  209.49  3750.0  251.13  3970.0000  -41.64  -220.0000
[40,]   69.00  1100.0   46.00   805.0000   23.00   295.0000
[41,]    0.00  2570.0    0.00  2593.7500    0.00   -23.7500
[42,] 5680.00 26075.0 3030.00 21525.0000 2650.00  4550.0000
[43,]  141.00  4700.0  194.00  8400.0000  -53.00 -3700.0000

In [24]:
dbar = rep(0,p)
for (i in (2*p+1):(3*p)){
  dbar[i-(2*p)] = sum(Y[,i])/n
}
Sd = matrix(0,p,p)
for (i in 1:n){
  Sd = Sd+1/(n-1)*((Y[i,(2*p+1):(3*p)]-dbar) %*% t(Y[i,(2*p+1):(3*p)]-dbar))
}
T2 = as.double(n*(t(dbar) %*% inv(Sd) %*% dbar))

cat("dbar = ",dbar,"\n")
cat("Sd = ",Sd,"\n")
cat("T2 = ",T2)

dbar =  117.7214 795.3287 
Sd =  176450 220507.1 220507.1 22352267 
T2 =  4.297374

Dari sini diperoleh
$$
\mathbf{\bar{d}} = \left(
\begin{array}{c}
     117.7214\\
     795.3287\\
\end{array}
\right),~~~~
\mathbf{S}_d = \left(
\begin{array}{cc}
     176450.0 &  220507.1\\
     220507.1 & 22352267.3
\end{array}
\right)
$$

Akan dilakukan uji hipotesis dengan
$$
      H_0 : \pmb{\mu}_d = \mathbf{0}
$$
dan $\alpha = 0.05$

Dengan menggunakan persamaan (5.24), diperoleh
\begin{align*}
T^2 &= n \mathbf{\bar{d}}' \mathbf{S}_d^{-1} \mathbf{\bar{d}}\\
&= (43) (117.7214,795.3287) \left(
\begin{array}{cc}
     176450.0 &  220507.1\\
     220507.1 & 22352267.3
\end{array}
\right)^{-1}
\left(
\begin{array}{c}
     117.7214\\
     795.3287\\
\end{array}
\right)\\
&= 4.297374
\end{align*}

Di sisi lain dari tabel A7 kita punya $T^2_{0.05,2,45} = 6.564$. Karena $T^2 = 4.297374 < 6.564$ maka $H_0$ tidak ditolak sehingga masih belum dapat disimpulkan **rata-rata EPS dan harga dari emiten-emiten pada index LQ45 tahun 2019 dan 2022 berbeda secara signifikan**


Selain dengan cara diatas, dapat juga digunakan fakta bahwa distribusi $T^2$ dapat didekati dengan distribusi $F$ yaitu
$$
\frac{(n-p)}{p(n-1)}T^2_{p,n-1} \sim  F_{p,n-p}
$$

In [25]:
F.T2= (n-p)/(p*(n-1))*T2
cat("F = ",F.T2,"\n")
cat("qf = ",qf(0.95,2,41))

F =  2.097528 
qf =  3.225684

Diperoleh hasil transformasi $T^2$ yang kita punya menjadi $F = 2.097528$ dan nilai dari $F_{0.05,2,41} = 3.225684$ sehingga diperoleh $F<F_{0.05,2,41}$ yang artinya $H_0$ tidak ditolak sesuai hasil yang telah diperoleh sebelumnya.

Selanjutnya akan dilihat apakah pandemi berpengaruh terhadap rata-rata EPS dan harga penutupan saham-saham pada sektor perbankan sebelum pandemi dan setelah pandemi. Akan digunakan data EPS emiten-emiten pada indeks LQ45 pada tahun 2019 dan 2022, dan harga penutupan pada tanggal 30 Desember 2019 dan 30 Desember 2022.

Untuk kemudahan dimisalkan
\begin{align*}
    y_1, x_1 &= \text{EPS tahunan pada tahun 2022, dan 2019}\\
    y_2, x_2 &= \text{Harga pada akhir tahun 2022, dan 2019}
\end{align*}

Untuk membandingkan efek pandemi dan sebelum pandemi pada awalnya dicari $d_1$ dan $d_2$ yang berturut-turut merupakan selisih dari $y_1,x_1$ dan $y_2,x_2$.


In [28]:
github_link <- "https://github.com/alzimna/Portfolio/raw/refs/heads/main/Test%20on%20Two%20Means/bank.xlsx"
temp_file <- tempfile(fileext = ".xlsx")
req <- GET(github_link,
          # authenticate using GITHUB_PAT
           authenticate(Sys.getenv("GITHUB_PAT"), ""),
          # write result to disk
           write_disk(path = temp_file))
Y = readxl::read_excel(temp_file)[,2:5]
Y

EPS 2022 PRICE 2022 EPS 2019 PRICE 2019
1   12.43   1765.0      -1.82    125.0000 
2    2.72     71.0      -3.69     61.0000 
3  231.00   2790.0     182.00   4250.0000 
4  384.00   2650.0     327.00   3250.0000 
5   13.79    925.0      18.13    322.0000 
6    1.61    131.0       2.25    300.0000 
7  330.00   8550.0     231.80   6685.0000 
8    3.59     80.0       4.75    129.0000 
9  202.21   1185.0     146.21    965.0000 
10 337.88   2730.0     416.78   3950.0000 
11   2.80     87.0       1.01     66.0000 
12   4.76     90.0     -29.69    135.0000 
13  26.45   3990.0       1.26    860.0000 
14   1.15   3720.0     -12.62   3100.0000 
15 -73.07    102.0      13.56    224.0000 
16 441.26   4962.5     294.45   3837.5000 
17  23.99   1090.0       8.59    358.0000 
18   2.20    520.0      29.47   9100.0000 
19  19.30    228.0      24.18    206.0000 
20 345.00   5275.0     165.13   6350.0000 
21 129.90   2120.0      60.53   2800.0000 
22   1.87    101.0       0.88     50.0000 
23  20.97    550.0       9.59    890.0000 
24 983.00   9225.0     825.00   7850.0000 
25 -83.75    645.0       2.68    284.0000 
26 144.99    745.0     128.10    845.0000 
27   5.99   1375.0      15.13   1750.0000 
28   0.93    170.0      -4.12    236.0000 
29 126.28   1540.0     137.70   1335.0000 
30   6.45     63.0       0.55     50.0000 
31  -4.61     50.0      -4.55    117.0528 
32 219.02   1345.0     157.98   1185.0000 
33 102.75    710.0      91.80    685.0000 
34  56.00   1015.0      47.20   1265.0000 
35 -24.28    102.0       0.32    180.0000 
36 338.00   4940.0     281.00   4400.0000 
37   0.50    404.0       2.34    198.0000 
38  11.22    845.0       0.43    585.0000 
39 102.54   1290.0       7.43    330.0000 
40 288.00   1350.0      17.77   2120.0000 
41  19.17    107.0      -1.56     84.0000 
42 100.44    570.0      75.11    830.0000

In [29]:
n = dim(Y)[1]
p = dim(Y)[2]/2
colnames(Y) = c("y1","y2","x1","x2")
Y$d1 = Y$y1-Y$x1
Y$d2 = Y$y2-Y$x2
Y = as.matrix(Y)
Y

y1     y2     x1     x2        d1     d2         
 [1,]  12.43 1765.0  -1.82  125.0000  14.25  1640.00000
 [2,]   2.72   71.0  -3.69   61.0000   6.41    10.00000
 [3,] 231.00 2790.0 182.00 4250.0000  49.00 -1460.00000
 [4,] 384.00 2650.0 327.00 3250.0000  57.00  -600.00000
 [5,]  13.79  925.0  18.13  322.0000  -4.34   603.00000
 [6,]   1.61  131.0   2.25  300.0000  -0.64  -169.00000
 [7,] 330.00 8550.0 231.80 6685.0000  98.20  1865.00000
 [8,]   3.59   80.0   4.75  129.0000  -1.16   -49.00000
 [9,] 202.21 1185.0 146.21  965.0000  56.00   220.00000
[10,] 337.88 2730.0 416.78 3950.0000 -78.90 -1220.00000
[11,]   2.80   87.0   1.01   66.0000   1.79    21.00000
[12,]   4.76   90.0 -29.69  135.0000  34.45   -45.00000
[13,]  26.45 3990.0   1.26  860.0000  25.19  3130.00000
[14,]   1.15 3720.0 -12.62 3100.0000  13.77   620.00000
[15,] -73.07  102.0  13.56  224.0000 -86.63  -122.00000
[16,] 441.26 4962.5 294.45 3837.5000 146.81  1125.00000
[17,]  23.99 1090.0   8.59  358.0000  15.40   732.00000
[18,]   2.20  520.0  29.47 9100.0000 -27.27 -8580.00000
[19,]  19.30  228.0  24.18  206.0000  -4.88    22.00000
[20,] 345.00 5275.0 165.13 6350.0000 179.87 -1075.00000
[21,] 129.90 2120.0  60.53 2800.0000  69.37  -680.00000
[22,]   1.87  101.0   0.88   50.0000   0.99    51.00000
[23,]  20.97  550.0   9.59  890.0000  11.38  -340.00000
[24,] 983.00 9225.0 825.00 7850.0000 158.00  1375.00000
[25,] -83.75  645.0   2.68  284.0000 -86.43   361.00000
[26,] 144.99  745.0 128.10  845.0000  16.89  -100.00000
[27,]   5.99 1375.0  15.13 1750.0000  -9.14  -375.00000
[28,]   0.93  170.0  -4.12  236.0000   5.05   -66.00000
[29,] 126.28 1540.0 137.70 1335.0000 -11.42   205.00000
[30,]   6.45   63.0   0.55   50.0000   5.90    13.00000
[31,]  -4.61   50.0  -4.55  117.0528  -0.06   -67.05279
[32,] 219.02 1345.0 157.98 1185.0000  61.04   160.00000
[33,] 102.75  710.0  91.80  685.0000  10.95    25.00000
[34,]  56.00 1015.0  47.20 1265.0000   8.80  -250.00000
[35,] -24.28  102.0   0.32  180.0000 -24.60   -78.00000
[36,] 338.00 4940.0 281.00 4400.0000  57.00   540.00000
[37,]   0.50  404.0   2.34  198.0000  -1.84   206.00000
[38,]  11.22  845.0   0.43  585.0000  10.79   260.00000
[39,] 102.54 1290.0   7.43  330.0000  95.11   960.00000
[40,] 288.00 1350.0  17.77 2120.0000 270.23  -770.00000
[41,]  19.17  107.0  -1.56   84.0000  20.73    23.00000
[42,] 100.44  570.0  75.11  830.0000  25.33  -260.00000

In [30]:
dbar = rep(0,p)
for (i in (2*p+1):(3*p)){
  dbar[i-(2*p)] = sum(Y[,i])/n
}
Sd = matrix(0,p,p)
for (i in 1:n){
  Sd = Sd+1/(n-1)*((Y[i,(2*p+1):(3*p)]-dbar) %*% t(Y[i,(2*p+1):(3*p)]-dbar))
}
T2 = as.double(n*(t(dbar) %*% inv(Sd) %*% dbar))
cat("dbar = ",dbar,"\n")
cat("Sd = ",Sd,"\n")
cat("T2 = ",T2)

dbar =  28.295 -50.92983 
Sd =  4504.846 15491.63 15491.63 2491143 
T2 =  7.842616

Dari sini diperoleh
$$
\mathbf{\bar{d}} = \left(
\begin{array}{c}
     28.29500\\
     -50.92983
\end{array}
\right),~~~~
\mathbf{S}_d = \left(
\begin{array}{cc}
     4504.846  & 15491.63\\
     15491.634 & 2491142.79
\end{array}
\right)
$$

Akan dilakukan uji hipotesis dengan
$$
    H_0 : \pmb{\mu}_d = \mathbf{0}.
$$
dan $\alpha = 0.05$

Dengan menggunakan persamaan (5.24), diperoleh
\begin{align*}
T^2 &= n \mathbf{\bar{d}}' \mathbf{S}_d^{-1} \mathbf{\bar{d}}\\
&= (42) (28.29500,-50.92983) \left(
\begin{array}{cc}
     4504.846  & 15491.63\\
     15491.634 & 2491142.79
\end{array}
\right)^{-1}
\left(
\begin{array}{c}
     28.29500\\
     -50.92983
\end{array}
\right)\\
&= 7.842616
\end{align*}


Di sisi lain dari tabel A7 kita punya $T^2_{0.05,2,45} = 6.564$. Karena $T^2 = 7.842616 > 6.564$ maka $H_0$ ditolak sehingga dapat disimpulkan **rata-rata EPS dan harga dari emiten-emiten pada sektor perbankan tahun 2019 dan 2022 berbeda secara signifikan.**

Selain dengan cara diatas, dapat juga digunakan fakta bahwa distribusi $T^2$ dapat didekati dengan distribusi $F$ yaitu
$$
\frac{(n-p)}{p(n-1)}T^2_{p,n-1} \sim  F_{p,n-p}
$$

In [31]:
F.T2= (n-p)/(p*(n-1))*T2
cat("F = ",F.T2,"\n")
cat("qf = ",qf(0.95,p,n-p))

F =  3.825666 
qf =  3.231727

Diperoleh hasil transformasi $T^2$ yang kita punya menjadi $F = 3.825666$ dan nilai dari $F_{0.05,2,41} = 3.231727$ sehingga diperoleh $F>F_{0.05,2,41}$ yang artinya $H_0$ ditolak sesuai hasil yang telah diperoleh sebelumnya.